### PR TITLE
[hail] add contig_recoding to import_bed and import_locus_intervals

### PR DESCRIPTION
**Pull request by danking**

### CODE REVIEW

File: hail/python/hail/methods/impex.py [hunk 1]
```diff
@@ -587,6 +596,14 @@ def import_locus_intervals(path, reference_genome='default', skip_invalid_interv
         Interval-keyed table.
     """
 
+    if contig_recoding is not None:
+        contig_recoding = hl.literal(contig_recoding)
+
+    def recode_contig(x):
+        if contig_recoding is None:
+            return x
+        return contig_recoding[x]
```
Comment (jigold) [inline hail/python/hail/methods/impex.py]: This should be the equivalent of `.get`. The dictionary of recordings doesn't have to have all contigs.

File: hail/python/hail/methods/impex.py [hunk 2]
```diff
@@ -752,8 +778,16 @@ def import_bed(path, reference_genome='default', skip_invalid_intervals=False, *
                               r"""^\w+=("[\w\d ]+"|\d+).*"""],
                      **kwargs)
 
+    if contig_recoding is not None:
+        contig_recoding = hl.literal(contig_recoding)
+
+    def recode_contig(x):
+        if contig_recoding is None:
+            return x
+        return contig_recoding[x]
```
Comment (jigold) [inline hail/python/hail/methods/impex.py]: See above.

File: hail/python/test/hail/methods/test_impex.py [hunk 3]
```diff
@@ -1223,6 +1223,15 @@ def test_import_locus_intervals_no_reference_specified(self):
...
         self.assertEqual(t.interval.dtype.point_type, hl.tstruct(contig=hl.tstr, position=hl.tint32))
 
+    def test_import_locus_intervals_recoding(self):
+        interval_file = resource('annotinterall.grch38.no.chr.interval_list')
+        t = hl.import_locus_intervals(interval_file,
+                                      contig_recoding={str(i): f'chr{i}' for i in [*range(22), 'X', 'Y', 'MT']},
+                                      reference_genome='GRCh38')
+        t._force_count()
+        self.assertTrue(t.count() == 2)
+        self.assertEqual(t.interval.dtype.point_type, hl.tstruct(contig=hl.tstr, position=hl.tint32))
```
Comment (jigold) [inline hail/python/test/hail/methods/test_impex.py]: Shouldn't this type be GRCh38 locus?

File: hail/python/test/hail/methods/test_impex.py [hunk 4]
```diff
@@ -1223,6 +1223,15 @@ def test_import_locus_intervals_no_reference_specified(self):
         self.assertTrue(t.count() == 2)
         self.assertEqual(t.interval.dtype.point_type, hl.tstruct(contig=hl.tstr, position=hl.tint32))
 
+    def test_import_locus_intervals_recoding(self):
+        interval_file = resource('annotinterall.grch38.no.chr.interval_list')
+        t = hl.import_locus_intervals(interval_file,
+                                      contig_recoding={str(i): f'chr{i}' for i in [*range(22), 'X', 'Y', 'MT']},
```
Comment (jigold) [inline hail/python/test/hail/methods/test_impex.py]: I'd take out MT here. I'm not sure chrMT is a thing in GRCh38.

File: hail/python/test/hail/methods/test_impex.py [hunk 5]
```diff
@@ -1263,6 +1272,25 @@ def test_import_bed(self):
...
+        bed = hl.import_bed(bed_file,
+                            reference_genome='GRCh38',
+                            contig_recoding={str(i): f'chr{i}' for i in [*range(22), 'X', 'Y', 'MT']})
+        bed._force_count()
+        nbed = bed.count()
+        i = 0
+        with open(bed_file) as f:
+            for line in f:
+                if len(line.strip()) != 0:
+                    try:
```
Comment (jigold) [inline hail/python/test/hail/methods/test_impex.py]: Is this just a complicated way to count the lines in the file? I like your tests above better.

File: hail/python/hail/methods/impex.py (hunk 1 above)
Comment (danking) [reply hail/python/hail/methods/impex.py]: Good catch, I added tests of both cases where the file has a mix of `chr22` and `22`.

File: hail/python/test/hail/methods/test_impex.py (hunk 3 above)
Comment (danking) [reply hail/python/test/hail/methods/test_impex.py]: Fixed.

File: hail/python/test/hail/methods/test_impex.py (hunk 4 above)
Comment (danking) [reply hail/python/test/hail/methods/test_impex.py]: Changed to `M`

File: hail/python/test/hail/methods/test_impex.py (hunk 5 above)
Comment (danking) [reply hail/python/test/hail/methods/test_impex.py]: I copied this from a test you added a year ago above ;). I hardcoded the expected count and added the type check.

File: hail/python/test/hail/methods/test_impex.py [hunk 6]
```diff
@@ -1263,6 +1272,14 @@ def test_import_bed(self):
 
         self.assertEqual(t.interval.collect(), hl.eval(expected))
 
+    def test_import_bed_recoding(self):
+        bed_file = resource('some-missing-chr-grch38.bed')
+        bed = hl.import_bed(bed_file,
+                            reference_genome='GRCh38',
+                            contig_recoding={str(i): f'chr{i}' for i in [*range(1, 23), 'X', 'Y', 'M']})
+        self.assertEqual(bed._force_count(), 3)
```
Comment (jigold) [inline hail/python/test/hail/methods/test_impex.py]: Shouldn't this be 5?

File: hail/python/test/hail/methods/test_impex.py [hunk 7]
```diff
@@ -1223,6 +1223,15 @@ def test_import_locus_intervals_no_reference_specified(self):
         self.assertTrue(t.count() == 2)
         self.assertEqual(t.interval.dtype.point_type, hl.tstruct(contig=hl.tstr, position=hl.tint32))
 
+    def test_import_locus_intervals_recoding(self):
+        interval_file = resource('annotinterall.grch38.no.chr.interval_list')
+        t = hl.import_locus_intervals(interval_file,
+                                      contig_recoding={str(i): f'chr{i}' for i in [*range(1, 23), 'X', 'Y', 'M']},
+                                      reference_genome='GRCh38')
+        t._force_count()
+        self.assertTrue(t.count() == 3)
```
Comment (jigold) [inline hail/python/test/hail/methods/test_impex.py]: You can combine these two lines as you did below.

Comment (tpoterba) [reply hail/python/test/hail/methods/test_impex.py]: also, don't use assertTrue. use assertEqual, but standard python `assert t.count() == 3` is preferred with pytest

Comment (danking) [reply hail/python/test/hail/methods/test_impex.py]: I removed every occurrence of `assertTrue(... == ...)` from this file.

**Pull request merged**